### DISCUSSION TITLE
pyvista running a docker image with Jupyter lab results in a blank image

### DISCUSSION THREAD

**Discussion by Balrog16** [pyvista/pyvista#6978] [Q&A]
### First check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the PyVista documentation, with the integrated search.
- [X] I already searched in Google "How to X in PyVista" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I will NOT make duplicate discussions or issues in the future.

### Commit to Help

- [X] I commit to help with one of those options.

### Sample Code What is the problem, question, or error?
Write a short description telling me what you are doing, what you expect to happen, and what is currently happening.

* used pip install to install pyvista
* uninstall vtk
* use requirements.txt to add other dependencies from 
 https://github.com/pyvista/pyvista/blob/main/docker/requirements.txt.

Yet when I try to run an example, say 

```python
import pyvista as pv
sphere = pv.Sphere()

# short example
sphere.plot(jupyter_backend='trame')

# long example
plotter = pv.Plotter(notebook=True)
plotter.add_mesh(sphere)
plotter.show(jupyter_backend='trame')

```
```


### Description

* It returns a blank image with just a outline
* Setting pv.set_jupyter_backend("static") makes it work (of course renders a static image)


### System Information

```shell
--------------------------------------------------------------------------------
  Date: Sat Dec 14 06:56:17 2024 UTC

                OS : Linux (Ubuntu 24.04)
            CPU(s) : 8
           Machine : x86_64
      Architecture : 64bit
               RAM : 7.7 GiB
       Environment : Jupyter
        GPU Vendor : Mesa/X.org
      GPU Renderer : llvmpipe (LLVM 15.0.6, 256 bits)
       GPU Version : 4.5 (Core Profile) Mesa 22.3.2
  MathText Support : True

  Python 3.12.3 (main, Sep 11 2024, 14:17:37) [GCC 13.2.0]

           pyvista : 0.44.1
               vtk : 9.3.1
             numpy : 2.0.2
        matplotlib : 3.9.2
            scooby : 0.10.0
             pooch : 1.8.2
            pillow : 10.4.0
           IPython : 8.28.0
        ipywidgets : 8.1.5
        jupyterlab : 4.2.5
             trame : 3.7.1
      trame_client : 3.5.0
      trame_server : 3.2.3
         trame_vtk : 2.8.12
     trame_vuetify : 2.7.2
      nest_asyncio : 1.6.0
--------------------------------------------------------------------------------
```

**Comment by Balrog16** [comment]
Though I don't understand much, setting pyvista.set_jupyter_backend('html') worked too! For example,

```python
import pyvista
pyvista.set_jupyter_backend('html')
from pyvista import examples
mesh = examples.download_st_helens()
warped = mesh.warp_by_scalar('Elevation')
surf = warped.extract_surface().triangulate()
surf = surf.decimate_pro(0.75)  # reduce the density of the mesh by 75%
surf.plot(cmap='gist_earth')
```

But this functionality is still limited unlike what one can see with "Trame"

**Comment by jourdain** [comment]
Unless you use vtk 9.4, the rendering will not work inside docker.

**Reply by Balrog16** [reply]
Thank you @jourdain, I find that the version of vtk used in the pyVista Docker is also 9.3!! Further, I do see renderings with backend set as html or static and they disappear only with backend set as Trame.

**Reply by jourdain** [reply]
The docker images from PyVista will work because they use a custom build of vtk with OSMesa. 
But if you simply install the default vtk, it will require X and expect a GPU, hence the issue you are describing.

**Comment by Balrog16** [comment]
I have now upgraded vtk to 9.4. And now there is a new message and the image is still blank, The message I think is exactly what you said. So, I installed `pip install --no-cache-dir --extra-index-url https://wheels.vtk.org vtk-osmesa`, but this message persists!

`vtkXOpenGLRenderWindow.:1290  WARN| vtkXOpenGLRenderWindow (0x52e0bf0): bad X server connection. DISPLAY=`

```python
from pyvista import examples
mesh = examples.download_st_helens()
warped = mesh.warp_by_scalar('Elevation')
surf = warped.extract_surface().triangulate()
surf = surf.decimate_pro(0.75)  # reduce the density of the mesh by 75%
surf.plot(cmap='gist_earth')
```

If I set the back end to `static` or `html`, the image is rendered!

--------------------------------------------------------------------------------
  Date: Tue Dec 17 14:49:20 2024 UTC

                  OS : Linux (Ubuntu 24.04)
              CPU(s) : 8
             Machine : x86_64
        Architecture : 64bit
                 RAM : 7.7 GiB
         Environment : Python
          GPU Vendor : Mesa/X.org
        GPU Renderer : llvmpipe (LLVM 15.0.6, 256 bits)
         GPU Version : 4.5 (Core Profile) Mesa 22.3.2
    MathText Support : True

  Python 3.12.3 (main, Sep 11 2024, 14:17:37) [GCC 13.2.0]

             pyvista : 0.44.1
                 vtk : 9.4.0
               numpy : 2.0.2
          matplotlib : 3.9.2
              scooby : 0.10.0
               pooch : 1.8.2
              pillow : 10.4.0
             imageio : 2.36.1
             IPython : 8.28.0
            colorcet : 3.1.0
             cmocean : 4.0.3
          ipywidgets : 8.1.5
              meshio : 5.3.5
          jupyterlab : 4.2.5
               trame : 3.7.1
        trame_client : 3.5.0
        trame_server : 3.2.3
           trame_vtk : 2.8.12
       trame_vuetify : 2.7.2
jupyter_server_proxy : 4.4.0
        nest_asyncio : 1.6.0
--------------------------------------------------------------------------------
--------------------------------------------------------------------------------

**Reply by jourdain** [reply]
If you want to go with `vtk-osmesa` you need to uninstall `vtk` first. How did you setup your docker? 

BTW vtk>=9.4 will warn that it can not use X, but should fallback to either EGL or OSMesa automatically. 

Maybe the white rendering, is not a rendering issue, but a network one? Hard to say from where I stand.

**Reply by Balrog16** [reply]
I am using the docker image provided by FEniCS -  [ghcr.io/fenics/dolfinx/lab:stable]( ghcr.io/fenics/dolfinx/lab:stable)
This comes with PyVista and the vtk version is 9.3. I have updated this with requirements.txt obtained from PyVista docker.

I am curious and also don't understand what you mean by "it is a network issue"? Any pointers in that direction please?

**Reply by user27182** [reply]
> I have now upgraded vtk to 9.4.

This may be a lost cause. vtk 9.4 is not yet supported by PyVista.

**Reply by jourdain** [reply]
@banesullivan did push something to reset the override but I'm not sure if that is inside a release or a dev branch. 

If things are below 9.4, then vtk needs to be manually uninstalled and then vtk-osmesa installed.

For the network part, the browser dev tool will be your friend to inspect what could be happening. But without screenshot or detailed explanation of what you are doing, I can only provide that [link](https://kitware.github.io/trame/guide/jupyter/intro.html) and go other the Jupyter section.
